File: Frontend/app/src/main/java/com/fezze/justus/data/repository/ApiRepository.kt
package com.fezze.justus.data.repository

import android.content.Context
import android.net.Uri
import com.fezze.justus.JustusApp
import com.fezze.justus.data.models.*
import com.fezze.justus.data.network.RetrofitClient
import com.fezze.justus.utils.ResultWrapper
import com.fezze.justus.utils.safeCallResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okio.BufferedSink

class ApiRepository {
    private val api = RetrofitClient.getApi(JustusApp.appContext)
    // -------------------- Auth / User --------------------
    suspend fun loginUser(usernameWithCode: String, password: String, deviceToken: String
    ): ResultWrapper<LoginResponse> = safeCallResult {
        api.login(LoginRequestBody(usernameWithCode, password, deviceToken))
    }
    suspend fun registerUser(username: String, password: String, email: String?, deviceToken: String
    ): ResultWrapper<RegisterResponse> = safeCallResult {
        api.register(RegisterRequestBody(username, password, email, deviceToken))
    }
    suspend fun requestUserCodes(username: String, password: String): ResultWrapper<RequestCodesResponse> =
        safeCallResult {
            api.requestUserCodes(RequestCodesBody(username, password))
        }
    suspend fun updateDeviceToken(usernameWithCode: String, deviceToken: String): ResultWrapper<Unit> =
        safeCallResult { api.updateDeviceToken(UpdateTokenRequest(usernameWithCode, deviceToken)) }
            .let { result ->
                when (result) {
                    is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                    is ResultWrapper.GenericError -> result
                    is ResultWrapper.NetworkError -> result
                }
            }
    // -------------------- Profile --------------------
    suspend fun fetchProfile(): ResultWrapper<User> =
        safeCallResult { api.getProfile() }.let { result ->
            when (result) {
                is ResultWrapper.Success ->
                    result.value.profile?.let { ResultWrapper.Success(it) }
                        ?: ResultWrapper.GenericError(null, "Profilo non trovato")
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun fetchPartnerProfile(): ResultWrapper<User> =
        safeCallResult { api.getPartnerProfile() }.let { result ->
            when (result) {
                is ResultWrapper.Success ->
                    result.value.profile?.let { ResultWrapper.Success(it) }
                        ?: ResultWrapper.GenericError(null, "Profilo partner non trovato")
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun updateProfileBio(bio: String?): ResultWrapper<Unit> =
        safeCallResult { api.updateProfile(UpdateProfileRequest(bio)) }
            .let { result ->
                when (result) {
                    is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                    is ResultWrapper.GenericError -> result
                    is ResultWrapper.NetworkError -> result
                }
            }
    suspend fun changePassword(oldPassword: String, newPassword: String): ResultWrapper<Unit> =
        safeCallResult {
            api.changePassword(ChangePasswordRequest(oldPassword, newPassword))
        }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    // -------------------- Partnership --------------------
    suspend fun sendPartnerRequest(partnerUsername: String, partnerCode: String): ResultWrapper<Unit> =
        safeCallResult { api.sendPartnerRequest(PartnerRequestBody(partnerUsername, partnerCode)) }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun acceptPartnerRequest(requesterId: Int): ResultWrapper<Unit> =
        safeCallResult { api.acceptPartnerRequest(AcceptRequestBody(requesterId)) }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun rejectPartnerRequest(requesterId: Int): ResultWrapper<Unit> =
        safeCallResult { api.rejectPartnerRequest(AcceptRequestBody(requesterId)) }.let { result ->
            when(result) {
                is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun searchPartner(username: String?, code: String?): ResultWrapper<List<User>> =
        safeCallResult { api.searchPartner(username, code) }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(result.value.users)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun getPartnership(): ResultWrapper<PartnershipResponse> =
        safeCallResult { api.getPartnership() }
    // -------------------- MissYou / Mood --------------------
    suspend fun sendMissYou(): ResultWrapper<Int> =
        safeCallResult { api.sendMissYou() }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(result.value.total)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun fetchMissYouTotal(): ResultWrapper<Int> =
        safeCallResult { api.getMissYouTotal() }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(result.value.total)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun setMood(emoji: String): ResultWrapper<MoodResponse> =
        safeCallResult { api.setMood(MoodRequest(emoji)) }
    suspend fun fetchMyMood(): ResultWrapper<MoodResponse> =
        safeCallResult { api.getMood("me") }
    suspend fun fetchPartnerMood(): ResultWrapper<MoodResponse> =
        safeCallResult { api.getMood("partner") }
    suspend fun fetchRecentCoupleEmojis(): ResultWrapper<List<String>> =
        safeCallResult { api.getRecentGlobalEmojis() }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(result.value.emojis)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    // -------------------- BucketList --------------------
    suspend fun fetchBucketList(): ResultWrapper<BucketListResponse> =
        safeCallResult { api.getBucket() }
    suspend fun addBucketItem(text: String): ResultWrapper<GenericResponse> =
        safeCallResult { api.addBucketItem(AddBucketRequest(text)) }
    suspend fun toggleBucketDone(id: Int): ResultWrapper<GenericResponse> =
        safeCallResult { api.toggleDone(id) }
    suspend fun deleteBucketItem(id: Int): ResultWrapper<GenericResponse> =
        safeCallResult { api.deleteBucketItem(id) }
    // -------------------- Game --------------------
    suspend fun fetchNewGameQuestion(): ResultWrapper<GameNewQuestionResponse> =
        safeCallResult {
            api.getNewQuestion()
        }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(result.value)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun submitGameAnswer(questionId: Int, votedFor: String): ResultWrapper<Unit> =
        safeCallResult {
            api.submitAnswer(GameAnswerRequest(questionId, votedFor))
        }.let { result ->
            when (result) {
                is ResultWrapper.Success -> ResultWrapper.Success(Unit)
                is ResultWrapper.GenericError -> result
                is ResultWrapper.NetworkError -> result
            }
        }
    suspend fun fetchGameStats(): ResultWrapper<GameStatsResponse> =
        safeCallResult { api.getGameStats() }
    // -------------------- Drive / Upload --------------------
    suspend fun fetchDriveItems(): ResultWrapper<DriveListResponse> =
        safeCallResult { api.getDriveItems() }
    suspend fun createDriveItem(type: String, content: String?, metadata: Map<String, Any>?): ResultWrapper<DriveSingleResponse> =
        safeCallResult { api.addDriveItem(DriveItemRequest(type, content, metadata)) }
    suspend fun deleteDriveItem(id: Int): ResultWrapper<GenericResponse> =
        safeCallResult { api.deleteDriveItem(id) }
    suspend fun fetchDriveChanges(lastSync: String?): ResultWrapper<DriveSyncResponse> =
        safeCallResult { api.getDriveChanges(lastSync) }
    // -------------------- Reactions --------------------
    suspend fun addReaction(itemId: Int, emoji: String): ResultWrapper<DriveItemReactionResponse> =
        safeCallResult {
            api.addDriveReaction(itemId, DriveItemReaction(emoji))
        }
    suspend fun fetchReactions(itemId: Int): ResultWrapper<DriveItemReactionsListResponse> =
        safeCallResult {
            api.getDriveReactions(itemId)
        }
    // -------------------- Favorite --------------------
    suspend fun addFavorite(itemId: Int): ResultWrapper<GenericResponse> =
        safeCallResult {
            api.addFavorite(itemId)
        }
    suspend fun removeFavorite(itemId: Int): ResultWrapper<GenericResponse> =
        safeCallResult {
            api.removeFavorite(itemId)
        }
    // -------------------- App Version --------------------
    suspend fun checkAppVersion(): ResultWrapper<AppVersionResponse> =
        safeCallResult { api.getAppVersion() }
    // -------------------- Upload --------------------
    suspend fun uploadProfile(context: Context, fileUri: Uri, fileName: String, mimeType: String,
                              onProgress: (uploadedBytes: Long) -> Unit): ResultWrapper<String> {
        return uploadFile(context, fileUri, fileName, mimeType, "/upload/profile", onProgress)
    }
    suspend fun uploadDiary(context: Context, fileUri: Uri, fileName: String, mimeType: String,
                            onProgress: (uploadedBytes: Long) -> Unit): ResultWrapper<String> {
        return uploadFile(context, fileUri, fileName, mimeType, "/upload/diary", onProgress)
    }
    private suspend fun uploadFile(context: Context, fileUri: Uri, fileName: String, mimeType: String, route: String,
                                   onProgress: (uploadedBytes: Long) -> Unit): ResultWrapper<String> {
        return withContext(Dispatchers.IO) {
            try {
                val filePart = prepareFilePartWithProgress(context, fileUri, fileName, mimeType, onProgress)
                val response = api.uploadFile(route, filePart)
                if (response.isSuccessful && response.body()?.success == true) {
                    ResultWrapper.Success(response.body()!!.url)
                } else {
                    ResultWrapper.GenericError(response.code(), "Upload fallito")
                }
            } catch (_: Exception) {
                ResultWrapper.NetworkError
            }
        }
    }
    fun prepareFilePartWithProgress(context: Context, fileUri: Uri, fileName: String, mimeType: String,
                                    onProgress: (uploadedBytes: Long) -> Unit): MultipartBody.Part {
        val inputStream = context.contentResolver.openInputStream(fileUri)!!
        val fileBytes = inputStream.readBytes()
        inputStream.close()
        val requestBody = object : RequestBody() {
            override fun contentType() = mimeType.toMediaTypeOrNull()
            override fun contentLength() = fileBytes.size.toLong()
            override fun writeTo(sink: BufferedSink) {
                var bytesWritten = 0L
                val buffer = ByteArray(32 * 1024)
                val input = fileBytes.inputStream()
                var read: Int
                while (input.read(buffer).also { read = it } != -1) {
                    sink.write(buffer, 0, read)
                    bytesWritten += read
                    onProgress(bytesWritten)
                }
                input.close()
            }
        }
        return MultipartBody.Part.createFormData("file", fileName, requestBody)
    }
}

File: Flutter_app/lib/repositories/api_repository.dart
// =============================================================================
// ApiRepository - Repository layer for API calls
// Dart equivalent of ApiRepository.kt
// =============================================================================

import '../models/models.dart';
import '../services/api_service.dart';
import '../services/result_wrapper.dart';

class ApiRepository {
  final ApiService _api = ApiService();

  // -------------------- Auth / User --------------------

  Future<ResultWrapper<LoginResponse>> loginUser(
    String usernameWithCode,
    String password,
    String deviceToken,
  ) async {
    return _api.login(LoginRequestBody(
      usernameWithCode: usernameWithCode,
      password: password,
      deviceToken: deviceToken,
    ));
  }

  Future<ResultWrapper<RegisterResponse>> registerUser(
    String username,
    String password,
    String? email,
    String deviceToken,
  ) async {
    return _api.register(RegisterRequestBody(
      username: username,
      password: password,
      email: email,
      deviceToken: deviceToken,
    ));
  }

  Future<ResultWrapper<RequestCodesResponse>> requestUserCodes(
    String username,
    String password,
  ) async {
    return _api.requestUserCodes(RequestCodesBody(
      username: username,
      password: password,
    ));
  }

  Future<ResultWrapper<void>> updateDeviceToken(
    String usernameWithCode,
    String deviceToken,
  ) async {
    final result = await _api.updateDeviceToken(UpdateTokenRequest(
      usernameWithCode: usernameWithCode,
      deviceToken: deviceToken,
    ));

    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  // -------------------- Profile --------------------

  Future<ResultWrapper<User>> fetchProfile() async {
    final result = await _api.getProfile();
    return switch (result) {
      Success(:final value) => value.profile != null
          ? Success(value.profile!)
          : const GenericError(message: 'Profilo non trovato'),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<User>> fetchPartnerProfile() async {
    final result = await _api.getPartnerProfile();
    return switch (result) {
      Success(:final value) => value.profile != null
          ? Success(value.profile!)
          : const GenericError(message: 'Profilo partner non trovato'),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<void>> updateProfileBio(String? bio) async {
    final result = await _api.updateProfile(UpdateProfileRequest(bio: bio));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<void>> changePassword(
    String oldPassword,
    String newPassword,
  ) async {
    final result = await _api.changePassword(ChangePasswordRequest(
      oldPassword: oldPassword,
      newPassword: newPassword,
    ));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  // -------------------- Partnership --------------------

  Future<ResultWrapper<void>> sendPartnerRequest(
    String partnerUsername,
    String partnerCode,
  ) async {
    final result = await _api.sendPartnerRequest(PartnerRequestBody(
      partnerUsername: partnerUsername,
      partnerCode: partnerCode,
    ));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<void>> acceptPartnerRequest(int requesterId) async {
    final result = await _api.acceptPartnerRequest(AcceptRequestBody(requesterId: requesterId));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<void>> rejectPartnerRequest(int requesterId) async {
    final result = await _api.rejectPartnerRequest(AcceptRequestBody(requesterId: requesterId));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<List<User>>> searchPartner(String? username, String? code) async {
    final result = await _api.searchPartner(username: username, code: code);
    return switch (result) {
      Success(:final value) => Success(value.users),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<PartnershipResponse>> getPartnership() async {
    return _api.getPartnership();
  }

  // -------------------- MissYou / Mood --------------------

  Future<ResultWrapper<int>> sendMissYou() async {
    final result = await _api.sendMissYou();
    return switch (result) {
      Success(:final value) => Success(value.total),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<int>> fetchMissYouTotal() async {
    final result = await _api.getMissYouTotal();
    return switch (result) {
      Success(:final value) => Success(value.total),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<MoodResponse>> setMood(String emoji) async {
    return _api.setMood(MoodRequest(emoji: emoji));
  }

  Future<ResultWrapper<MoodResponse>> fetchMyMood() async {
    return _api.getMood('me');
  }

  Future<ResultWrapper<MoodResponse>> fetchPartnerMood() async {
    return _api.getMood('partner');
  }

  Future<ResultWrapper<List<String>>> fetchRecentCoupleEmojis() async {
    final result = await _api.getRecentGlobalEmojis();
    return switch (result) {
      Success(:final value) => Success(value.emojis),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  // -------------------- BucketList --------------------

  Future<ResultWrapper<BucketListResponse>> fetchBucketList() async {
    return _api.getBucket();
  }

  Future<ResultWrapper<GenericResponse>> addBucketItem(String text) async {
    return _api.addBucketItem(AddBucketRequest(text: text));
  }

  Future<ResultWrapper<GenericResponse>> toggleBucketDone(int id) async {
    return _api.toggleBucketDone(id);
  }

  Future<ResultWrapper<GenericResponse>> deleteBucketItem(int id) async {
    return _api.deleteBucketItem(id);
  }

  // -------------------- Game --------------------

  Future<ResultWrapper<GameNewQuestionResponse>> fetchNewGameQuestion() async {
    return _api.getNewQuestion();
  }

  Future<ResultWrapper<void>> submitGameAnswer(int questionId, String votedFor) async {
    final result = await _api.submitAnswer(GameAnswerRequest(
      questionId: questionId,
      votedFor: votedFor,
    ));
    return switch (result) {
      Success() => const Success(null),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<GameStatsResponse>> fetchGameStats() async {
    return _api.getGameStats();
  }

  // -------------------- Drive / Upload --------------------

  Future<ResultWrapper<DriveListResponse>> fetchDriveItems() async {
    return _api.getDriveItems();
  }

  Future<ResultWrapper<DriveSingleResponse>> createDriveItem(
    String type,
    String? content,
    Map<String, dynamic>? metadata,
  ) async {
    return _api.addDriveItem(DriveItemRequest(
      type: type,
      content: content,
      metadata: metadata,
    ));
  }

  Future<ResultWrapper<GenericResponse>> deleteDriveItem(int id) async {
    return _api.deleteDriveItem(id);
  }

  Future<ResultWrapper<DriveSyncResponse>> fetchDriveChanges(String? lastSync) async {
    return _api.getDriveChanges(lastSync: lastSync);
  }

  // -------------------- Reactions --------------------

  Future<ResultWrapper<DriveItemReactionResponse>> addReaction(int itemId, String emoji) async {
    return _api.addDriveReaction(itemId, DriveItemReaction(emoji: emoji));
  }

  Future<ResultWrapper<DriveItemReactionsListResponse>> fetchReactions(int itemId) async {
    return _api.getDriveReactions(itemId);
  }

  // -------------------- Favorite --------------------

  Future<ResultWrapper<GenericResponse>> addFavorite(int itemId) async {
    return _api.addFavorite(itemId);
  }

  Future<ResultWrapper<GenericResponse>> removeFavorite(int itemId) async {
    return _api.removeFavorite(itemId);
  }

  // -------------------- Upload --------------------

  Future<ResultWrapper<String>> uploadProfile(
    List<int> fileBytes,
    String fileName,
    String mimeType,
  ) async {
    final result = await _api.uploadFile(
      route: '/upload/profile',
      fileBytes: fileBytes,
      fileName: fileName,
      mimeType: mimeType,
    );
    return switch (result) {
      Success(:final value) => Success(value.url),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  Future<ResultWrapper<String>> uploadDiary(
    List<int> fileBytes,
    String fileName,
    String mimeType,
  ) async {
    final result = await _api.uploadFile(
      route: '/upload/diary',
      fileBytes: fileBytes,
      fileName: fileName,
      mimeType: mimeType,
    );
    return switch (result) {
      Success(:final value) => Success(value.url),
      GenericError(:final code, :final message) => GenericError(code: code, message: message),
      NetworkError() => const NetworkError(),
    };
  }

  // -------------------- App Version --------------------

  Future<ResultWrapper<AppVersionResponse>> checkAppVersion() async {
    return _api.getAppVersion();
  }

  void dispose() {
    _api.dispose();
  }
}
